api/V1/monitoringPage.ts

```diff
@@ -0,0 +1,182 @@
+    from? : Date;
+    until? : Date;
+    page : number;
+    pageSize: number;
+}
+
+export interface MonitoringPageCriteria {
+    compareAi: string;
+    devices? : number[];
+    groups? : number[];
+    page: number,
+    pagesEstimate: number,
+    pageFrom?: Date;
+    pageUntil?: Date;
+    searchFrom?: Date;
+    searchUntil?: Date;
+}
+
+
+export async function calculateMonitoringPageCriteria(params : MonitoringParams, viewAsSuperAdmin: boolean) : Promise<MonitoringPageCriteria> {
+    return getDatesForSearch(params, viewAsSuperAdmin);
+}
+
+async function getDatesForSearch(params: MonitoringParams, viewAsSuperAdmin: boolean) : Promise<MonitoringPageCriteria> {
+    
+    const replacements = {
+        GROUPS_AND_DEVICES : makeGroupsAndDevicesCriteria(params.devices, params.groups),
+        USER_PERMISSIONS: await makeGroupsAndDevicesPermissions(params.user, viewAsSuperAdmin),
+        DATE_SELECTION: makeDatesCriteria(params),
+        PAGING: null
+    }
+
+    const countRet = await models.sequelize.query(replaceInSQL(VISITS_COUNT_SQL, replacements), { type: QueryTypes.SELECT });
+    const approxVisitCount = parseInt(countRet[0].count);
+    const returnVal = createPageCriteria(params, approxVisitCount);
+
+    if (approxVisitCount < params.pageSize) {
+        returnVal.pageFrom = returnVal.searchFrom;
+        returnVal.pageUntil = returnVal.searchUntil;        
+    } else if (params.page <= returnVal.pagesEstimate ) {
+        const limit : number = params.pageSize * 1 + 1;
+        const offset : number = (params.page - 1) * params.pageSize;
+        replacements.PAGING = ` LIMIT ${ limit } OFFSET ${ offset }`;
+        const results = await models.sequelize.query(replaceInSQL(VISIT_STARTS_SQL, replacements), { type: QueryTypes.SELECT });
+
+        if (results.length > 0) {
+            returnVal.pageUntil = (params.page == 1) ? returnVal.searchUntil : results[0].recordingDateTime;
+            if (params.page < returnVal.pagesEstimate) {
+                returnVal.pageFrom = results[results.length - 1].recordingDateTime;
+            } else {
+                returnVal.pageFrom = returnVal.searchFrom;
+            }
+        }
+    } 
+
+    return returnVal;
+}
+
+function createPageCriteria(params: MonitoringParams, count: number) : MonitoringPageCriteria {
+    const criteria : MonitoringPageCriteria = {
+        page: params.page,
+        pagesEstimate: Math.ceil(count / params.pageSize),
+        searchFrom: params.from || BEFORE_CACOPHONY, 
+        searchUntil: params.until || new Date(), 
+        compareAi: "Master"
+    }
+
+    if (params.devices) {
+        criteria.devices = params.devices;
+    }
+
+    if (params.groups) {
+        criteria.groups = params.groups;
+    }
+
+    return criteria;
+}
+
+function replaceInSQL(sql: string, replacements: { [key: string]: string} ) : string {
+    for (const key in replacements) {
+        sql = sql.replace(`{${key}}`, replacements[key]);
```

Comment:
This seems suspect - key is a string, but you're using string interpolation on it by itself.  Why?
Also, string replace as used will only replace the first instance of the string found, which doesn't seem to be what you want.  You need to use a regex with the global flag set: `new RegExp(key, 'g')` : `https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace`